src/Microsoft.Azure.SignalR/Startup/NegotiateMatcherPolicy.cs

```diff
@@ -0,0 +1,75 @@
+﻿// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#if NETCOREAPP3_0
+using System.Collections.Concurrent;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Http.Connections;
+using Microsoft.AspNetCore.Routing;
+using Microsoft.AspNetCore.Routing.Matching;
+using Microsoft.AspNetCore.SignalR;
+
+namespace Microsoft.Azure.SignalR
+{
+    internal class NegotiateMatcherPolicy : MatcherPolicy, IEndpointSelectorPolicy
+    {
+        // This caches the replacement endpoints for negotiate so they are not recomputed on every request
+        private readonly ConcurrentDictionary<Endpoint, Endpoint> _negotiateEndpointCache = new ConcurrentDictionary<Endpoint, Endpoint>();
+        
+        public override int Order => 1;
+
+        public bool AppliesToEndpoints(IReadOnlyList<Endpoint> endpoints)
+        {
+            foreach (var endpoint in endpoints)
+            {
+                var hubMetadata = endpoint.Metadata.GetMetadata<HubMetadata>();
+                var negotiateMetadata = endpoint.Metadata.GetMetadata<NegotiateMetadata>();
+
+                if (hubMetadata != null && negotiateMetadata != null)
+                {
+                    return true;
+                }
+            }
+
+            return false;
+        }
+
+        public Task ApplyAsync(HttpContext httpContext, EndpointSelectorContext context, CandidateSet candidates)
+        {
+            for (var i = 0; i < candidates.Count; i++)
+            {
+                ref var candidate = ref candidates[i];
+                var newEndpoint = _negotiateEndpointCache.GetOrAdd(candidate.Endpoint, CreateNegotiateEndpoint);
+
+                candidates.ReplaceEndpoint(i, newEndpoint, candidate.Values);
+            }
+
+            return Task.CompletedTask;
+        }
+
+        private Endpoint CreateNegotiateEndpoint(Endpoint endpoint)
+        {
+            var routeEndpoint = (RouteEndpoint)endpoint;
+            var hubMetadata = endpoint.Metadata.GetMetadata<HubMetadata>();
+
+            // Replaces the negotiate endpoint with one that does the service redirect
+            var routeEndpointBuilder = new RouteEndpointBuilder(async context =>
```

Comment:
We should be setting the `DisplayName` for this endpoint.

Reply:
Thank you @BrennanConroy for helping review. Is there any naming convention we should follow ? Is **RouteAzureSignalRService** good?

Reply:
The norm seems to be the route path, SignalR sets it to `"/<hub path>/negotiate"` and `"/<hub path>"` for the two endpoints